pkg/tfbridge/validate_input_types.go

```diff
@@ -0,0 +1,471 @@
+
+	// The path to the property that failed
+	ResourcePath string
+}
+
+// NewInputValidator creates a new input validator for a given resource and
+// package schema
+func NewInputValidator(urn resource.URN, schema pschema.PackageSpec) *PulumiInputValidator {
+	return &PulumiInputValidator{
+		urn:    urn,
+		schema: schema,
+	}
+}
+
+// getArrayTypeSpec gets a type spec for an array type. Arrays can have their
+// type defined in a couple different ways. This will check the different ways
+// and return the correct type spec If the array supports multiple types (i.e.
+// `oneOf`) then it will find the correct one and return it. This
+// can be used to provide a more helpful error message
+func (v *PulumiInputValidator) getArrayTypeSpec(
+	spec *pschema.TypeSpec,
+	propertyValue resource.PropertyValue,
+) (
+	typeSpec *pschema.TypeSpec,
+	requiredProps []string,
+) {
+	if spec.Items != nil {
+		typeSpec = spec.Items
+	}
+	if len(spec.OneOf) > 0 {
+		typeSpec, _, _ = v.findOneOf(propertyValue, spec.OneOf)
+	}
+
+	return typeSpec, requiredProps
+}
+
+// getObjectTypeSpec gets a type spec for an object type. Objects can have their
+// type defined in a couple different ways. This will check the different ways
+// and return the correct type spec.
+func (v *PulumiInputValidator) getObjectTypeSpec(
+	spec *pschema.TypeSpec,
+	propertyValue resource.PropertyValue,
+) (
+	complexSpec *pschema.ComplexTypeSpec,
+	typeSpec *pschema.TypeSpec,
+	requiredProps []string,
+) {
+
+	if spec.AdditionalProperties != nil {
+		if spec.AdditionalProperties.Ref != "" {
+			complexSpec = v.getType(spec.AdditionalProperties.Ref)
+			// if we don't find the reference then just continue instead of failing
+			if complexSpec != nil {
+				requiredProps = complexSpec.Required
+			}
+		} else if len(spec.AdditionalProperties.OneOf) > 0 {
+			typeSpec, _, _ = v.findOneOf(propertyValue, spec.OneOf)
+		} else {
+			typeSpec = spec.AdditionalProperties
+		}
+	} else if spec.Ref != "" {
+		complexSpec = v.getType(spec.Ref)
+		if complexSpec != nil {
+			requiredProps = complexSpec.Required
+		}
+	} else if len(spec.OneOf) > 0 {
+		typeSpec, _, _ = v.findOneOf(propertyValue, spec.OneOf)
+		// If the OneOf that we found is an object with a `Type` then return that sub type
+		// Otherwise we should just return the OneOf type that we found
+		if typeSpec.AdditionalProperties != nil && typeSpec.AdditionalProperties.Type != "" {
+			typeSpec = typeSpec.AdditionalProperties
+		}
+	}
+	return complexSpec, typeSpec, requiredProps
+}
+
+// validateTypeSpec is the main function for validating a PropertyValue against the pulumi schema.
+// It is a recursive function that will validate nested types and arrays. It will return a list of
+// type failures if any are found
+func (v *PulumiInputValidator) validateTypeSpec(
+	propertyKey string,
+	propertyValue resource.PropertyValue,
+	typeSpec pschema.TypeSpec,
+	pathBuilder pathBuilder,
+	requiredProps []string,
+) *[]TypeFailure {
```

Comment:
Why do we need `*[]TypeFailure`? Go slices are already nullable.
```suggestion
) []TypeFailure {
```